File: src/main/java/pl/mmakos/advent/year2024/Day1.java
package pl.mmakos.advent.year2024;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import pl.mmakos.advent.utils.Pair;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.IntStream;

import static java.lang.Math.abs;
import static pl.mmakos.advent.utils.Utils.*;

@NoArgsConstructor(access = AccessLevel.NONE)
public final class Day1 {
  @SuppressWarnings("java:S106")
  public static void main(String[] args) {
    System.out.println("Advent of code 2024, day 1");
    System.out.println("TASK 1: " + task1());
    System.out.println("TASK 2: " + task2());
  }

  private static Pair<List<Integer>, List<Integer>> getInput() {
    List<Integer> first = new ArrayList<>();
    List<Integer> second = new ArrayList<>();
    lines()
        .map(l -> splitToInts(l, "\\s+"))
        .forEach(ints -> {
          first.add(ints[0]);
          second.add(ints[1]);
        });
    return new Pair<>(first, second);
  }

  private static int task1() {
    Pair<List<Integer>, List<Integer>> input = getInput();
    input.first().sort(Integer::compareTo);
    input.second().sort(Integer::compareTo);
    return IntStream.range(0, input.first().size())
        .map(i -> abs(input.first().get(i) - input.second().get(i)))
        .sum();
  }

  private static long task2() {
    Pair<List<Integer>, List<Integer>> input = getInput();
    return input.first().stream()
        .mapToLong(f -> f * input.second().stream()
            .filter(s -> Objects.equals(s, f))
            .count())
        .sum();
  }
}

File: src/main/cpp/year2024/day1.cpp
#include <iostream>
#include <vector>

#include <ranges>
#include <utils.h>
#include <algorithm>

constexpr int day = 1;
constexpr int year = 2024;

std::pair<std::vector<int>, std::vector<int>> parseInput(const std::vector<std::string> &input) {
  std::vector<int> first;
  std::vector<int> second;
  const std::regex re(R"(\s+)");
  for (const auto &s : input) {
    auto split = aoc::split(s, re);
    first.push_back(std::stoi(split[0]));
    second.push_back(std::stoi(split[1]));
  }
  return std::make_pair(first, second);
}


int solve1(const std::vector<std::string> &input) {
  auto [first, second] = parseInput(input);
  std::ranges::sort(first);
  std::ranges::sort(second);
  int sum = 0;
  for (const auto [a, b] : std::views::zip(first, second)) {
    sum += std::abs(a - b);
  }
  return sum;
}

long long solve2(const std::vector<std::string> &input) {
  auto [first, second] = parseInput(input);

  long long sum = 0;
  for (const auto a : first) {
    sum += a * std::ranges::count(second, a);
  }
  return sum;
}

int main() {
  const std::vector<std::string> input = aoc::readInput<std::string>(year, day);

  std::cout << "Advent of code " << year << ", day " << day << std::endl;
  std::cout << "TASK 1: " << solve1(input) << std::endl;
  std::cout << "TASK 2: " << solve2(input) << std::endl;
}
